File: vidaa-tv/js/m3u-parser.js
/**
 * 2TV Shared IPTV M3U Playlist Parser (Universal ES5/ES6 compatible)
 * Parses #EXTINF channel names, tvg-logo, group-title, and stream URLs.
 */
(function (window) {
    'use strict';

    var M3uParser = {
        isM3uPlaylist: function (urlOrPath, contentHead) {
            var lower = (urlOrPath || '').toLowerCase().trim();
            if (lower.indexOf('.m3u') !== -1 && lower.indexOf('.m3u8') === -1) return true;
            if (contentHead && contentHead.indexOf('#EXTINF:') !== -1 && contentHead.indexOf('#EXT-X-STREAM-INF') === -1) {
                return true;
            }
            return false;
        },

        parse: function (content) {
            var channels = [];
            if (!content) return channels;

            var lines = content.split(/\r?\n/);
            var currentName = '';
            var currentLogo = null;
            var currentGroup = null;

            for (var i = 0; i < lines.length; i++) {
                var line = lines[i].trim();

                if (line.indexOf('#EXTINF:') === 0 || line.indexOf('#extinf:') === 0) {
                    // Extract tvg-logo
                    var logoMatch = /tvg-logo="([^"]+)"/i.exec(line);
                    currentLogo = logoMatch ? logoMatch[1] : null;

                    // Extract group-title
                    var groupMatch = /group-title="([^"]+)"/i.exec(line);
                    currentGroup = groupMatch ? groupMatch[1] : null;

                    // Extract channel name (after comma)
                    var commaIndex = line.lastIndexOf(',');
                    if (commaIndex !== -1) {
                        currentName = line.substring(commaIndex + 1).trim();
                    } else {
                        currentName = 'Canale ' + (channels.length + 1);
                    }
                } else if (line.length > 0 && line.indexOf('#') !== 0) {
                    if (!currentName) {
                        currentName = 'Canale ' + (channels.length + 1);
                    }
                    channels.push({
                        name: currentName,
                        streamUrl: line,
                        logoUrl: currentLogo,
                        groupTitle: currentGroup
                    });
                    currentName = '';
                    currentLogo = null;
                    currentGroup = null;
                }
            }
            return channels;
        }
    };

    window.M3uParser = M3uParser;
})(window);

File: android/tv/src/main/java/com/twotv/tv/util/M3uParser.kt
package com.twotv.tv.util

data class M3uChannel(
    val name: String,
    val streamUrl: String,
    val logoUrl: String? = null,
    val groupTitle: String? = null
)

object M3uParser {
    fun isM3uPlaylist(urlOrPath: String, contentHead: String = ""): Boolean {
        val lower = urlOrPath.lowercase()
        return lower.endsWith(".m3u") || lower.contains(".m3u?") || contentHead.contains("#EXTM3U")
    }

    fun parse(content: String): List<M3uChannel> {
        val channels = mutableListOf<M3uChannel>()
        val lines = content.lines()
        var currentName = ""
        var currentLogo: String? = null
        var currentGroup: String? = null

        for (line in lines) {
            val trimmed = line.trim()
            if (trimmed.startsWith("#EXTINF:", ignoreCase = true)) {
                // Extract tvg-logo if present
                val logoMatch = Regex("""tvg-logo="([^"]+)"""", RegexOption.IGNORE_CASE).find(trimmed)
                currentLogo = logoMatch?.groupValues?.get(1)

                // Extract group-title if present
                val groupMatch = Regex("""group-title="([^"]+)"""", RegexOption.IGNORE_CASE).find(trimmed)
                currentGroup = groupMatch?.groupValues?.get(1)

                // Extract channel name (after comma)
                val commaIndex = trimmed.lastIndexOf(',')
                currentName = if (commaIndex != -1) {
                    trimmed.substring(commaIndex + 1).trim()
                } else {
                    "Canale ${channels.size + 1}"
                }
            } else if (trimmed.isNotBlank() && !trimmed.startsWith("#")) {
                if (currentName.isBlank()) {
                    currentName = "Canale ${channels.size + 1}"
                }
                channels.add(
                    M3uChannel(
                        name = currentName,
                        streamUrl = trimmed,
                        logoUrl = currentLogo,
                        groupTitle = currentGroup
                    )
                )
                currentName = ""
                currentLogo = null
                currentGroup = null
            }
        }
        return channels
    }
}
